core/services/ocr2/plugins/ocr2keeper/evm21/upkeepstate/orm.go

```diff
@@ -0,0 +1,89 @@
+package upkeepstate
+
+import (
+	"database/sql"
+	"errors"
+	"math/big"
+	"time"
+
+	"github.com/smartcontractkit/sqlx"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/pg"
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
+)
+
+type orm struct {
+	chainID *big.Int
+	q       pg.Q
+}
+
+type PersistedStateRecord struct {
+	UpkeepID            []byte
+	WorkID              string
+	CompletionState     uint8
+	BlockNumber         uint64
+	IneligibilityReason uint8
+	AddedAt             time.Time
+}
+
+// NewORM creates an ORM scoped to chainID.
+func NewORM(chainID *big.Int, db *sqlx.DB, lggr logger.Logger, cfg pg.QConfig) *orm {
+	return &orm{
+		chainID: chainID,
+		q:       pg.NewQ(db, lggr.Named("ORM"), cfg),
+	}
+}
+
+// InsertUpkeepState is idempotent and sets upkeep state values in db
+func (o *orm) InsertUpkeepState(state PersistedStateRecord, qopts ...pg.QOpt) error {
+	q := o.q.WithOpts(qopts...)
+
+	query := `INSERT INTO evm_upkeep_state (evm_chain_id, work_id, completion_state, block_number, added_at, upkeep_id, ineligibility_reason)
+	  VALUES ($1::NUMERIC, $2, $3, $4::NUMERIC, $5, $6::BYTEA, $7::NUMERIC)
+	    ON CONFLICT (evm_chain_id, work_id)
+	    DO NOTHING`
+
+	return q.ExecQ(query, utils.NewBig(o.chainID), state.WorkID, state.CompletionState, state.BlockNumber, state.AddedAt, state.UpkeepID, state.IneligibilityReason)
+}
+
+// SelectStatesByWorkIDs searches the data store for stored states for the
+// provided work ids and configured chain id
+func (o *orm) SelectStatesByWorkIDs(workIDs []string, qopts ...pg.QOpt) (states []PersistedStateRecord, err error) {
+	q := o.q.WithOpts(qopts...)
+
+	namedArgs := map[string]any{
+		"chainID": utils.NewBig(o.chainID),
+		"workIDs": workIDs,
+	}
+
+	query, args, err := sqlx.Named(`SELECT upkeep_id, work_id, completion_state, block_number, ineligibility_reason, added_at
+	  FROM evm_upkeep_state
+	  WHERE evm_chain_id = :chainID AND work_id IN (:workIDs)`, namedArgs)
```

Comment:
There is an easier way to do this.
```suggestion
	  WHERE evm_chain_id = :chainID AND work_id = ANY (:workIDs)`, namedArgs)
```

This way allows you to pass in the slice as a named parameter and negates the need for `sqlx.In`.

Reply:
Doing this causes a syntax error: `ERROR: syntax error at or near "AND" (SQLSTATE 42601)`. Maybe the order is important?

Reply:
`ANY (` should be `ANY(`

Reply:
You must also not use `In` any more you pass a `pq.Array(workIDs)`. Check other examples elsewhere in the code for guidance.